[libc][NFC] fix int warnings in float conversion

michaelrj-google opened the pull request:
The printf float to string conversion functions had some implicit
integer conversion warnings on gcc. This patch adds explicit casts to
these places.

michaelrj-google committed:
[libc][NFC] fix int warnings in float conversion

The printf float to string conversion functions had some implicit
integer conversion warnings on gcc. This patch adds explicit casts to
these places.

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -548,7 +550,8 @@ class FloatToString {
 #endif
-      const int32_t shift_amount = SHIFT_CONST + (-exponent - IDX_SIZE * idx);
+      const int32_t shift_amount =
+          static_cast<int32_t>(SHIFT_CONST + (-exponent - IDX_SIZE * idx));
```
Sorry, I think I merged this in https://github.com/llvm/llvm-project/pull/74369. I wonder what page I was looking at where I didn't observe warnings on the above?

michaelrj-google replied:
I've rebased to pick that up, and also matched the narrower cast scope you had.

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -157,9 +157,10 @@ LIBC_INLINE constexpr uint32_t ceil_log10_pow2(const uint32_t e) {
-         BLOCK_SIZE;
+  return static_cast<uint32_t>((ceil_log10_pow2(idx) +
+                                ceil_log10_pow2(mantissa_width + 1) +
+                                (BLOCK_SIZE - 1)) /
+                               BLOCK_SIZE);
```
I think just BLOCK_SIZE is a size_t which is resulting in promotions.  You could limit the static cast to just those (or perhaps define BLOCK_SIZE as a smaller type?)

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -488,7 +489,8 @@ class FloatToString {
 #endif
-      const uint32_t shift_amount = SHIFT_CONST + (IDX_SIZE * idx) - exponent;
+      const uint32_t shift_amount =
+          static_cast<uint32_t>(SHIFT_CONST + (IDX_SIZE * idx) - exponent);
```
I think just IDX_SIZE is a size_t; You could limit the static cast to just those (or perhaps define IDX_SIZE as a smaller type?)

nickdesaulniers commented on libc/src/math/generic/math_utils.h:
```diff
@@ -34,7 +34,9 @@ LIBC_INLINE double as_double(uint64_t x) { return cpp::bit_cast<double>(x); }
 LIBC_INLINE uint32_t top12_bits(float x) { return as_uint32_bits(x) >> 20; }
 
-LIBC_INLINE uint32_t top12_bits(double x) { return as_uint64_bits(x) >> 52; }
+LIBC_INLINE uint32_t top12_bits(double x) {
+  return static_cast<uint32_t>(as_uint64_bits(x) >> 52);
+}
```
does this function have any users in tree?

michaelrj-google committed:
Rebase and narrow scope of casts in float_to_string

Patch #74369 fixed one of the fixed in the previous commit. This rebases
on top of that commit and also adjusts the casts in that file to better
match.

nickdesaulniers commented on libc/src/stdio/printf_core/float_dec_converter.h:
```diff
@@ -864,7 +867,7 @@ LIBC_INLINE int convert_float_dec_auto_typed(Writer *writer,
         // If the next digit is not the same as the previous one, then there are
         // no more contiguous trailing digits.
-        if ((copy_of_digits % 10) != cur_last_digit) {
+        if (static_cast<int>(copy_of_digits % 10) != cur_last_digit) {
```
/home/llvm-libc-buildbot/buildbot-worker/libc-x86_64-debian-fullbuild/libc-x86_64-debian-gcc-fullbuild-dbg/llvm-project/libc/src/stdio/printf_core/float_dec_converter.h:867:35: error: comparison of integer expressions of different signedness: ‘__llvm_libc_18_0_0_git::BlockInt’ {aka ‘unsigned int’} and ‘int’ [-Werror=sign-compare]


if BlockInt is unsigned, perhaps copy_of_digits should also be unsigned, and modulo 10U?

michaelrj-google replied:
I think you mean `cur_last_digit` should be unsigned, which I agree with.

nickdesaulniers commented on libc/src/__support/UInt.h:
```diff
@@ -121,6 +121,7 @@ template <size_t Bits, bool Signed> struct BigInt {
         return lo;
       }
     } else {
+      // TODO: silence shift warning
```
I guess Bits could be a value like 65, hence the warning?  Maybe

```c
if constexpr (Bits == 32) {
  return static_cast<T>((static_cast<T>(val[1]) << 64) + lo);
} else {
  ...
```
would silence the warning?

nickdesaulniers replied:
then perhaps a pragma here disabling the warning is the best approach?

I guess the static_assert on line 30 would allow values like 128 or 256?

michaelrj-google replied:
I'm going to leave fixing this warning for a followup patch.

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -490,7 +490,7 @@ class FloatToString {
 #endif
       const uint32_t shift_amount =
-          static_cast<uint32_t>(SHIFT_CONST + (IDX_SIZE * idx) - exponent);
+          SHIFT_CONST + static_cast<uint32_t>(IDX_SIZE * idx) - exponent;
```
here `idx` is a `uint32_t`, so the cast only need be on IDX_SIZE

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -551,7 +551,7 @@ class FloatToString {
 #endif
       const int32_t shift_amount =
-          static_cast<int32_t>(SHIFT_CONST + (-exponent - IDX_SIZE * idx));
+          SHIFT_CONST + (-exponent - static_cast<int32_t>(IDX_SIZE * idx));
```
here, `idx` is a `int32_t`, so the cast only need be on `IDX_SIZE`

michaelrj-google committed:
more type tweaks

nickdesaulniers commented on libc/src/__support/float_to_string.h:
```diff
@@ -290,7 +290,7 @@ LIBC_INLINE cpp::UInt<MID_INT_SIZE> get_table_negative(int exponent, size_t i) {
       ten_blocks = 0;
       five_blocks = i;
-      shift_amount = static_cast<int>(shift_amount + (i * BLOCK_SIZE));
+      shift_amount = shift_amount + (static_cast<int>(i) * BLOCK_SIZE);
```
I think you can remove the parens here now; operator precedence

here and below, there seems to be excess parens?

michaelrj-google replied:
personally I prefer to have parentheses when an equation is ambiguous. Even if operator precedence does the correct thing, it's less cognitive load to clarify with parentheses.

nickdesaulniers commented on libc/src/stdio/printf_core/float_dec_converter.h:
```diff
@@ -537,7 +537,7 @@ LIBC_INLINE int convert_float_decimal_typed(Writer *writer,
 
   if (exponent < MANT_WIDTH) {
-    const uint32_t blocks = (precision / BLOCK_SIZE) + 1;
+    const uint32_t blocks = static_cast<uint32_t>(precision / BLOCK_SIZE) + 1;
```
you can maybe get more specific here and below; the static_cast can be applied to just BLOCK_SIZE I suspect?

nickdesaulniers reviewed (approved):
thanks for the changes. I don't feel strongly about the remaining comments; feel free to address them or not.

michaelrj-google committed:
final tweaks before landing
